napari/viewer.py

```diff
@@ -118,3 +118,15 @@ def close(self):
             # https://github.com/napari/napari/issues/1500
             for layer in self.layers:
                 chunk_loader.on_layer_deleted(layer)
+
+    @classmethod
+    def current(cls) -> Optional['Viewer']:
```

Comment:
I like how `Viewer.current()` gives the current viewer instance if any, but I'd find it a little odd if `viewer.current()` didn't give me back the exact same `viewer`, i.e `viewer.current() == viewer`. Now maybe classmethods just shouldn't be called on an instance, but I'm not sure it's strictly impossible, and so we might want to do something to prevent people from getting confused.

Alternatively I'd rather put this on the window, so it was more like `viewer.window.current_viewer()` or something. I know that `current` isn't on our `ViewerModel`, but I'm hoping one day for us to be able to do more passing of the `ViewerModel` to plugins and have the difference between the `ViewerModel` and the `Viewer` be as small as possible (maybe even none, with more of this app logic passed to the `Window` or another even higher level container which contains all this stuff, including lists of the viewers etc).

Thinking even more about that idea, in the long run I could see us with an API like

```python
viewer # instance of a viewer, same as now
viewer.window # same as is now, the instance of a window corresponding to the viewer
viewer.application # singleton application variable, there is only ever one of these. this would be new
viewer.application.viewers # selectable evented list of all the viewers that have been created
viewer.application.viewers[0] # instance of the first viewer in list
viewer.application.viewers.current # current viewer
```

@tlambert03 if you and others like this API proposal now we could think about it, or we could go with the current API and then move to that one later. We could also just do `Viewer._current()` for now if we want to keep our options open a little longer, I'm very open to ideas and see all this as improvements.

Thanks for the review too @goanpeca and @Carreau. Curious what you think of my API proposal too

Reply:
> I'd find it a little odd if viewer.current() didn't give me back the exact same viewer, i.e viewer.current() == viewer. Now maybe classmethods just shouldn't be called on an instance, but I'm not sure it's strictly impossible, and so we might want to do something to prevent people from getting confused.

Since `.current()` isn't a universal python API with prior meaning, can't the docstring just explain what it is?  It also seems like a bit of a funny expectation that someone would make an `obj.method()` that simply returned `obj` ... so I'm not sure how often someone would stumble on this and go "great! I have a method that gives me back the object I called it from!" :)  

While I think that this pattern fits among common usage of `classmethod`,  I do think I could make an alternate decorator that prevented it from being called on a viewer instance if people wanted that.

> Alternatively I'd rather put this on the window, so it was more like viewer.window.current_viewer()

I think I'd rather just remove it from viewer altogether... Feels weirder to me that `_qt.qt_main_window.Window` would be used to return an instance of `viewer.Viewer`. 

as for `ViewerModel`, I thought about it a little bit but I don't think this concept of "current viewer" belongs on the viewer model at all.  It's a very GUI-specific concept (in a headless mode, you really don't have an active/current instance... they are all equally "current" and accessible) ... and as the `Viewer` is the combination of our headless `ViewerModel` with a gui `Window` ... that's why I thought it belonged there.  
 
But, for the majority of our internal needs, `_QtMainWindow.current()` will actually be sufficient.  So, if you want to punt on `Viewer.current` altogether and leave it strictly internal, I'm fine with that too...


> ```
> viewer.application # singleton application variable, there is only ever one of these. this would be new
> viewer.application.viewers # selectable evented list of all the viewers that have been created
> viewer.application.viewers[0] #
> ```

meh :) 
honestly, I'm not sure we need to engineer this one too much with an evented model... (maybe?  but definitely not yet).

Reply:
since the `Viewer.current` classmethod isn't really necessary for things like the notification manager, and for #2282 ... (both of which can use `_QtMainWindow.current()`), I removed it from this PR so we can discuss it further before expanding the viewer api

Reply:
> Since .current() isn't a universal python API with prior meaning, can't the docstring just explain what it is? It also seems like a bit of a funny expectation that someone would make an obj.method() that simply returned obj ... so I'm not sure how often someone would stumble on this and go "great! I have a method that gives me back the object I called it from!" :)

Totally, agreed - that's a bit funny too!!

> since the Viewer.current classmethod isn't really necessary for things like the notification manager, and for #2282 ... (both of which can use _QtMainWindow.current()), I removed it from this PR so we can discuss it further before expanding the viewer api

Ok yeah, I'm def not 100% opposed to that, but waiting seems good here. I think we can get this in now without it